File: server/services/query-executor.ts
import { Pool, Client } from "pg";
import { DatabaseConfigService } from "./database-config";

export interface QueryResult {
  table_data: any[];
  chart_data: {
    type: 'bar' | 'line' | 'pie';
    data: any[];
    xAxis: string;
    yAxis: string;
  };
  columns: string[];
  row_count: number;
  execution_time: number;
}

export class QueryExecutor {
  private static instance: QueryExecutor;
  private connectionPools: Map<string, Pool> = new Map();
  
  static getInstance(): QueryExecutor {
    if (!QueryExecutor.instance) {
      QueryExecutor.instance = new QueryExecutor();
    }
    return QueryExecutor.instance;
  }

  private getConnectionPoolKey(connection: any): string {
    return `${connection.host}:${connection.port}/${connection.database}/${connection.username}`;
  }

  private getOrCreatePool(connection: any): Pool {
    const poolKey = this.getConnectionPoolKey(connection);
    
    if (this.connectionPools.has(poolKey)) {
      const existingPool = this.connectionPools.get(poolKey)!;
      // Check if pool is still valid
      if (existingPool.totalCount >= 0) {
        return existingPool;
      } else {
        // Pool was ended, remove it
        this.connectionPools.delete(poolKey);
      }
    }

    // Check if this is a cloud database that requires SSL
    const requiresSSL = connection.host.includes('neon.tech') || 
                       connection.host.includes('supabase.') ||
                       connection.host.includes('amazonaws.com') ||
                       connection.host.includes('planetscale.') ||
                       connection.host.includes('railway.');

    const pool = new Pool({
      host: connection.host,
      port: connection.port,
      database: connection.database,
      user: connection.username,
      password: connection.password,
      ssl: requiresSSL ? { rejectUnauthorized: false } : false,
      connectionTimeoutMillis: 30000,
      idleTimeoutMillis: 30000,
      max: 5, // Maximum number of connections in pool
      allowExitOnIdle: true,
    });

    // Handle pool errors
    pool.on('error', (err) => {
      console.error('Pool error:', err);
      this.connectionPools.delete(poolKey);
    });

    this.connectionPools.set(poolKey, pool);
    return pool;
  }

  async executeQuery(sqlQuery: string, userId: string): Promise<QueryResult> {
    const startTime = Date.now();
    const dbService = DatabaseConfigService.getInstance();
    const currentConnection = dbService.getCurrentConnection(userId);

    if (!currentConnection) {
      throw new Error("No active database connection");
    }

    const pool = this.getOrCreatePool(currentConnection);
    let client;

    try {
      client = await pool.connect();
      
      // Sanitize SQL query - extract and validate SELECT statements
      let sanitizedQuery = sqlQuery.trim();
      
      // Remove code block formatting if present
      if (sanitizedQuery.startsWith('```sql') || sanitizedQuery.startsWith('```')) {
        const lines = sanitizedQuery.split('\n');
        const sqlLines = lines.slice(1, -1); // Remove first and last lines (```sql and ```)
        sanitizedQuery = sqlLines.join(' ').trim(); // Join with space instead of newline
      }
      
      // Remove any remaining markdown or extra formatting
      sanitizedQuery = sanitizedQuery.replace(/^```sql\s*/i, '').replace(/\s*```$/, '');
      
      // Ensure proper SQL formatting with correct spacing
      sanitizedQuery = sanitizedQuery
        .replace(/\s+/g, ' ') // Replace multiple spaces/newlines with single space
        .replace(/\bFROM\b/gi, ' FROM ')
        .replace(/\bWHERE\b/gi, ' WHERE ')
        .replace(/\bGROUP BY\b/gi, ' GROUP BY ')
        .replace(/\bORDER BY\b/gi, ' ORDER BY ')
        .replace(/\bHAVING\b/gi, ' HAVING ')
        .replace(/\bJOIN\b/gi, ' JOIN ')
        .replace(/\bLEFT JOIN\b/gi, ' LEFT JOIN ')
        .replace(/\bRIGHT JOIN\b/gi, ' RIGHT JOIN ')
        .replace(/\bINNER JOIN\b/gi, ' INNER JOIN ')
        .replace(/\s+/g, ' ') // Clean up any double spaces
        .trim();
      
      // Log the sanitized query for debugging
      console.log('Original SQL:', sqlQuery);
      console.log('Sanitized SQL:', sanitizedQuery);
      
      if (!sanitizedQuery.toLowerCase().startsWith('select')) {
        throw new Error(`Only SELECT queries are allowed for security reasons. Received: ${sanitizedQuery.substring(0, 100)}...`);
      }

      const result = await client.query(sanitizedQuery);
      
      // Handle null values in results more intelligently
      const cleanedRows = result.rows.map(row => {
        const cleanedRow: any = {};
        for (const [key, value] of Object.entries(row)) {
          // Only convert null to 0 for chart processing, preserve nulls for display
          if (value === null) {
            // For averages, counts, sums - preserve null to show "No data"
            if (key.toLowerCase().includes('avg') || key.toLowerCase().includes('average')) {
              cleanedRow[key] = null;
            } else {
              cleanedRow[key] = 0;
            }
          } else {
            cleanedRow[key] = value;
          }
        }
        return cleanedRow;
      });
      
      const executionTime = (Date.now() - startTime) / 1000;
      const columns = result.fields.map(field => field.name);

      // Generate chart data based on the cleaned results
      const chartData = this.generateChartData(cleanedRows, columns);

      return {
        table_data: cleanedRows,
        chart_data: chartData,
        columns: columns,
        row_count: cleanedRows.length,
        execution_time: executionTime
      };

    } catch (error) {
      console.error("Query execution error:", error);
      throw new Error(`Query execution failed: ${error instanceof Error ? error.message : String(error)}`);
    } finally {
      if (client) {
        client.release();
      }
    }
  }

  // Clean up all connection pools
  async cleanup(): Promise<void> {
    const cleanupPromises: Promise<void>[] = [];
    
    this.connectionPools.forEach((pool, key) => {
      cleanupPromises.push(
        pool.end().catch(err => {
          console.error(`Error closing pool ${key}:`, err);
        })
      );
    });
    
    await Promise.allSettled(cleanupPromises);
    this.connectionPools.clear();
  }

  private generateChartData(rows: any[], columns: string[]): {
    type: 'bar' | 'line' | 'pie';
    data: any[];
    xAxis: string;
    yAxis: string;
  } {
    if (rows.length === 0 || columns.length === 0) {
      return {
        type: 'bar',
        data: [],
        xAxis: '',
        yAxis: ''
      };
    }

    // Find numeric and text columns
    const numericColumns = columns.filter(col => {
      const sampleValue = rows[0]?.[col];
      return typeof sampleValue === 'number' || (!isNaN(parseFloat(sampleValue)) && isFinite(parseFloat(sampleValue)));
    });

    const textColumns = columns.filter(col => {
      const sampleValue = rows[0]?.[col];
      return typeof sampleValue === 'string' && isNaN(parseFloat(sampleValue));
    });

    const dateColumns = columns.filter(col => {
      const sampleValue = rows[0]?.[col];
      return sampleValue instanceof Date || (typeof sampleValue === 'string' && !isNaN(Date.parse(sampleValue)));
    });

    // Determine chart type and axes
    let chartType: 'bar' | 'line' | 'pie' = 'bar';
    let xAxis = columns[0];
    let yAxis = columns[1] || columns[0];

    // If we have date columns, prefer line chart
    if (dateColumns.length > 0 && numericColumns.length > 0) {
      chartType = 'line';
      xAxis = dateColumns[0];
      yAxis = numericColumns[0];
    }
    // If we have text and numeric columns, use bar chart
    else if (textColumns.length > 0 && numericColumns.length > 0) {
      chartType = 'bar';
      xAxis = textColumns[0];
      yAxis = numericColumns[0];
    }
    // If we have only a few rows with categories, use pie chart
    else if (rows.length <= 10 && textColumns.length > 0 && numericColumns.length > 0) {
      chartType = 'pie';
      xAxis = textColumns[0];
      yAxis = numericColumns[0];
    }

    // Process data for chart
    const chartData = rows.map(row => {
      const processedRow: any = {};
      
      // Convert numeric strings to numbers for better charting
      for (const [key, value] of Object.entries(row)) {
        if (numericColumns.includes(key) && typeof value === 'string') {
          processedRow[key] = parseFloat(value);
        } else {
          processedRow[key] = value;
        }
      }
      
      return processedRow;
    });

    return {
      type: chartType,
      data: chartData,
      xAxis,
      yAxis
    };
  }
}

File: backend/services/query_executor.py
from __future__ import annotations

import re
import time
from typing import Any, Dict, List

import psycopg

from .database_config import DatabaseConfigService


class QueryExecutor:
    _instance: QueryExecutor | None = None

    def __init__(self) -> None:
        pass

    @classmethod
    def get_instance(cls) -> QueryExecutor:
        if cls._instance is None:
            cls._instance = QueryExecutor()
        return cls._instance

    async def execute_query(self, sql_query: str, user_id: str) -> Dict[str, Any]:
        start = time.time()
        db_service = DatabaseConfigService.get_instance()
        current = db_service.get_current_connection(user_id)
        if not current:
            raise ValueError("No active database connection")

        # Sanitize SQL: allow only SELECT
        sanitized = sql_query.strip()
        # strip markdown code fences
        if sanitized.startswith("```"):
            sanitized = re.sub(r"^```(?:sql)?\s*|\s*```$", "", sanitized, flags=re.IGNORECASE)
        sanitized = re.sub(r"\s+", " ", sanitized)
        if not sanitized.lower().startswith("select"):
            raise ValueError("Only SELECT queries are allowed for security reasons.")

        requires_ssl = any(
            host in current.get("host", "")
            for host in ["neon.tech", "supabase.", "amazonaws.com", "planetscale.", "railway."]
        )
        conninfo = psycopg.conninfo.make_conninfo(
            host=current.get("host"),
            port=current.get("port"),
            dbname=current.get("database"),
            user=current.get("username"),
            password=current.get("password"),
            sslmode="require" if requires_ssl else "disable",
        )

        async with await psycopg.AsyncConnection.connect(conninfo) as aconn:
            async with aconn.cursor() as cur:
                await cur.execute(sanitized)
                rows = [dict(zip([col.name for col in cur.description], r)) for r in await cur.fetchall()]

        # Clean nulls and build columns
        cleaned_rows: List[Dict[str, Any]] = []
        for row in rows:
            new_row: Dict[str, Any] = {}
            for key, value in row.items():
                if value is None:
                    if "avg" in key.lower() or "average" in key.lower():
                        new_row[key] = None
                    else:
                        new_row[key] = 0
                else:
                    new_row[key] = value
            cleaned_rows.append(new_row)

        columns = list(rows[0].keys()) if rows else []
        chart = self._generate_chart_data(cleaned_rows, columns)

        execution_time = time.time() - start
        return {
            "table_data": cleaned_rows,
            "chart_data": chart,
            "columns": columns,
            "row_count": len(cleaned_rows),
            "execution_time": round(execution_time, 3),
        }

    def _generate_chart_data(self, rows: List[Dict[str, Any]], columns: List[str]) -> Dict[str, Any]:
        if not rows or not columns:
            return {"type": "bar", "data": [], "xAxis": "", "yAxis": ""}

        def is_number(v: Any) -> bool:
            if isinstance(v, (int, float)):
                return True
            try:
                float(v)
                return True
            except Exception:
                return False

        numeric_cols = [c for c in columns if is_number(rows[0].get(c))]
        text_cols = [c for c in columns if isinstance(rows[0].get(c), str) and not is_number(rows[0].get(c))]
        date_cols = [
            c
            for c in columns
            if isinstance(rows[0].get(c), (str,))
            and _is_date_like(rows[0].get(c))
        ]

        chart_type = "bar"
        x_axis = columns[0]
        y_axis = columns[1] if len(columns) > 1 else columns[0]

        if date_cols and numeric_cols:
            chart_type = "line"
            x_axis = date_cols[0]
            y_axis = numeric_cols[0]
        elif text_cols and numeric_cols:
            chart_type = "bar"
            x_axis = text_cols[0]
            y_axis = numeric_cols[0]
        elif len(rows) <= 10 and text_cols and numeric_cols:
            chart_type = "pie"
            x_axis = text_cols[0]
            y_axis = numeric_cols[0]

        chart_data: List[Dict[str, Any]] = []
        for row in rows:
            processed: Dict[str, Any] = {}
            for key, value in row.items():
                if key in numeric_cols and isinstance(value, str):
                    try:
                        processed[key] = float(value)
                    except Exception:
                        processed[key] = value
                else:
                    processed[key] = value
            chart_data.append(processed)

        return {"type": chart_type, "data": chart_data, "xAxis": x_axis, "yAxis": y_axis}


def _is_date_like(s: Any) -> bool:
    try:
        import dateutil.parser  # type: ignore
        dateutil.parser.parse(str(s))
        return True
    except Exception:
        try:
            import datetime as _dt
            _dt.datetime.fromisoformat(str(s))
            return True
        except Exception:
            return False


